File: day_03.jl
letras = ["a" ,"b" ,"c" ,"d" , "e", "f","g" , "h", "i" ,"j" , "k" , "l" , "m", "n", "o", "p", "q" , "r", "s", "t" , "u", "v", "w", "x", "y" , "z"]
letras_dict = Dict()
for (index, letra) in enumerate(letras)
    letras_dict[letra] = index
    letras_dict[uppercase(letra)] = index+26
end

# part-01
prioridades = []
io = open("input_03.txt", "r")
for row in eachline(io)
    m = eachmatch(r"\w+", row)
    matches = collect(m)
    rucksack = string(matches[1].match[:])
    largo = length(rucksack)
    mitad = Int64(largo/2)
    letra = intersect(collect(string(SubString(rucksack, 1, mitad))), collect(string(SubString(rucksack, mitad+1, largo))))[1]
    push!(prioridades,letras_dict[string(letra)])
end
println(sum(prioridades))

# part-02
filas = []
io = open("input_03.txt", "r")
for row in eachline(io)
    m = eachmatch(r"\w+", row)
    matches = collect(m)
    rucksack = string(matches[1].match[:])
    push!(filas,rucksack)
end
prioridades = []
for idx in 1:3:length(filas)-2
    letra = intersect(collect(string(filas[idx])), collect(string(filas[idx+1])), collect(string(filas[idx+2])))[1]
    push!(prioridades,letras_dict[string(letra)])
end
println(sum(prioridades))

File: day_03.clj
(require '[clojure.java.io :as io])

(def letras 
  ["a" ,"b" ,"c" ,"d" , "e", "f", "g" , "h", "i" ,"j" , "k" , "l" , "m", "n", "o", "p", "q" , "r", "s", "t" , "u", "v", "w", "x", "y" , "z"])
(def hash-letras 
  (loop [hash-temp {} idx 0]
    (if (< idx (* (count letras) 2))
      (if (< idx (count letras))
        (recur (assoc hash-temp (keyword (nth letras idx)) (inc idx)) (inc idx))
        (recur (assoc hash-temp (keyword (clojure.string/upper-case (nth letras (- idx 26)))) (inc idx)) (inc idx)))
      hash-temp)))


;; part-01
(with-open [rdr (io/reader "resources/input_03.txt")]
  (let [lines (line-seq rdr) num-rows (- (count lines) 1)]
    (loop [line (first lines) idx 0 prioridades []]
      (if (< idx num-rows)
        (let [[rucksack] (re-seq #"\w+" line)
              largo (count rucksack)
              mitad (/ largo 2)
              subs1 (subs rucksack 0 mitad)
              subs2 (subs rucksack mitad)]
          (recur (nth lines (inc idx)) (inc idx) (conj prioridades (get hash-letras (keyword (first (clojure.set/intersection (set (map str (set subs1))) (set (map str (set subs2))))))))))
        (let [[rucksack] (re-seq #"\w+" line)
              largo (count rucksack)
              mitad (/ largo 2)
              subs1 (subs rucksack 0 mitad)
              subs2 (subs rucksack mitad)]
          (reduce + (conj prioridades (get hash-letras (keyword (first (clojure.set/intersection (set (map str (set subs1))) (set (map str (set subs2))))))))))))))

;; part-02
(with-open [rdr (io/reader "resources/input_03.txt")]
  (let [lines (line-seq rdr) num-rows (- (count lines) 1)]
    (loop [line (first lines) idx 0 group [] prioridades []]
      (if (< idx num-rows)
        (if (< (count group) 3)
          (let [[rucksack] (re-seq #"\w+" line)]
            (recur (nth lines (inc idx)) (inc idx) (conj group (set (map str (set rucksack)))) prioridades))
          (let [[rucksack] (re-seq #"\w+" line)]
            (recur (nth lines (inc idx)) (inc idx) (conj [] (set (map str (set rucksack)))) (conj prioridades (get hash-letras (keyword (first (apply clojure.set/intersection group))))))))
        (let [[rucksack] (re-seq #"\w+" line)]
          (reduce + (conj prioridades (get hash-letras (keyword (first (apply clojure.set/intersection (conj group (set (map str (set rucksack)))))))))))))))